recipes-extended/quark/files/add-POST-and-v4v-support.patch

```diff
@@ -0,0 +1,319 @@
+POST support: very basic; added to support surf sending log messages via POST
+
+v4v support: added a few select() calls to prevent server lockup when serving
+resources
+
+################################################################################
+CHANGELOG
+################################################################################
+x64 patch: Nicholas Tsirakis <tsirakisn@ainfosec.com>
+
+################################################################################
+REMOVAL
+################################################################################
+
+################################################################################
+UPSTREAM PLAN
+################################################################################
+
+################################################################################
+INTERNAL DEPENDENCIES
+################################################################################
+
+################################################################################
+PATCHES
+################################################################################
+
+commit 8ea834685a163c7dd20ac6c46e0efcae43f8d2e8
+Author: Nicholas Tsirakis <tsirakisn@ainfosec.com>
+Date:   Tue Feb 12 10:22:24 2019 -0500
+
+    x64 patch
+
+diff --git a/config.mk b/config.mk
+index 7056241..668b0ef 100644
+--- a/config.mk
++++ b/config.mk
+@@ -4,13 +4,13 @@ VERSION = 0
+ # Customize below to fit your system
+ 
+ # paths
+-PREFIX = /usr/local
++PREFIX ?= /usr
+ MANPREFIX = $(PREFIX)/share/man
+ 
+ # flags
+-CPPFLAGS = -DVERSION=\"$(VERSION)\" -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=700 -D_BSD_SOURCE
+-CFLAGS   = -std=c99 -pedantic -Wall -Wextra -Os
+-LDFLAGS  = -s
++CPPFLAGS ?= -DVERSION=\"$(VERSION)\" -D_DEFAULT_SOURCE -D_XOPEN_SOURCE=700 -D_BSD_SOURCE
++CFLAGS   ?= -std=c99 -pedantic -Wall -Wextra -O
++LDFLAGS  ?=
+ 
+ # compiler and linker
+-CC = cc
++CC ?= gcc
+diff --git a/http.c b/http.c
+index c848e81..f53c25c 100644
+--- a/http.c
++++ b/http.c
+@@ -30,6 +30,7 @@ const char *req_field_str[] = {
+ const char *req_method_str[] = {
+ 	[M_GET]  = "GET",
+ 	[M_HEAD] = "HEAD",
++	[M_POST] = "POST"
+ };
+ 
+ const char *status_str[] = {
+@@ -121,14 +122,37 @@ http_get_request(int fd, struct request *r)
+ 		}
+ 	}
+ 
+-	/* remove terminating empty line */
+-	if (hlen < 2) {
++	/* split header and payload (if applicable)
++	 *
++	 *  expected POST format: HEADER,
++	 *                        two newlines,
++	 *                        PAYLOAD,
++	 *                        two newlines
++	 *
++	 *  expected GET format: HEADER,
++	 *                       two newlines
++	 */
++
++	/* find first occurrence of double newlines */
++	p = strstr(h, "\r\n\r\n");
++	if (!p) {
+ 		return http_send_status(fd, S_BAD_REQUEST);
+ 	}
+-	hlen -= 2;
+ 
+-	/* null-terminate the header */
+-	h[hlen] = '\0';
++	/* keep the first newline */
++	p += 2;
++
++	/* find second occurrence - if exists, POST request, else GET */
```

Comment:
Assuming this is reading the full HTTP message, POST or GET should be the absolute first characters seen, so it would be far more efficient to look in the absolute positions than a string match when handling GET requests. The first occurrence of double-CRLF is required, but the second search could probably be taken out completely.

Reply:
The second search was less about figuring out the type of request and more about finding where to null terminate it and copying the correct number of bytes from the payload. That being said the second search can be done conditionally to avoid searching unnecessarily for a GET request.